File: gred/src/flag.lua
Flag = class()
Flag._ActiveTexture = nil
Flag._InactiveTexture = nil

function Flag:Create(id, x, y, z)
  self = Flag:New()
  self.id = id
  self.entity = CreateSprite(nil, MATERIAL_ALPHA)
  self.active = false
  self:position(x, y, z)
  self:setActive(true)
  return self
end

function Flag:destroy()
  FreeEntity(self.entity)
end

function Flag:position(x, y, z)
  SetEntityPosition(self.entity, x, y, z)
end

function Flag:x()
  return EntityX(self.entity)
end

function Flag:y()
  return EntityY(self.entity)
end

function Flag:z()
  return EntityZ(self.entity)
end

function Flag:setActive(active)
  if self._ActiveTexture == nil then self._ActiveTexture = LoadTexture("icons/flag_red.png") end
  if self._InactiveTexture == nil then self._InactiveTexture = LoadTexture("icons/flag_orange.png") end
  if active ~= self.active then
    self.active = active
    if active then
      SetMaterialTexture(EntityMaterial(self.entity, 1), 1, self._ActiveTexture)
    else
      SetMaterialTexture(EntityMaterial(self.entity, 1), 1, self._InactiveTexture)
    end
  end
end

File: src/flag.cpp
#include "flag.h"

Texture* Flag::active_texture   = nullptr;
Texture* Flag::inactive_texture = nullptr;

Flag::Flag(int id_, float x, float y, float z) : id(id_) {
    if (!active_texture)
        active_texture   = CacheTexture(App::resource_path("icons/flag_red.png").c_str(),   FILTER_NONE);
    if (!inactive_texture)
        inactive_texture = CacheTexture(App::resource_path("icons/flag_orange.png").c_str(), FILTER_NONE);

    entity = CreateSprite(inactive_texture, 1, 1, SPRITE_SPHERICAL, nullptr);
    SetSpriteSize(entity, 1.f, 1.f);
    Material* mat = GetEntityMaterial((Entity*)entity, 0);
    SetMaterialBlendMode(mat, BLEND_ALPHA);
    SetMaterialLightingEnabled(mat, FALSE);
    active = false;
    position(x, y, z);
    set_active(true);
}

void Flag::destroy() {
    FreeEntity((Entity*)entity);
}

void Flag::position(float x, float y, float z) {
    App::set_entity_position((Entity*)entity, x, y, z);
}

float Flag::x() const { return App::entity_x((Entity*)entity); }
float Flag::y() const { return App::entity_y((Entity*)entity); }
float Flag::z() const { return App::entity_z((Entity*)entity); }

void Flag::set_active(bool a) {
    if (a != active) {
        active = a;
        Texture* tex = active ? active_texture : inactive_texture;
        SetMaterialColorTexture(GetEntityMaterial((Entity*)entity, 0), tex);
    }
}
